### DISCUSSION TITLE
Check module compatibility at container startup: ✅⚠️🟠❌

### DISCUSSION THREAD

**Discussion by felddy** [felddy/foundryvtt-docker#991] [Show and tell]
The container uses a startup flag called `--noupdate` that prevents updates from being applied through the GUI, and notifies the user to `pull` a new image.  A side-effect of this is that the module compatibility checker is inaccessible, as it is tied to the update process.

As a work-around I've created a container patch that will do a compatibility check when you startup the container.

- https://gist.github.com/felddy/a02d66c3817f8223652d96febe580153

You can have it run at container startup by adding the URL to the `CONTAINER_PATCH_URLS` environment variable:

```yaml
      - CONTAINER_PATCH_URLS=
        https://gist.githubusercontent.com/felddy/a02d66c3817f8223652d96febe580153/raw/foundryvtt-module-check.sh
```
>[!warning]
> Only use patch URLs from trusted sources!

>[!note]
> Alternatively you can download the script and place in a directory inside your `data` mount.  Then set the `CONTAINER_PATCHES` environment variable to point to the directory.  e.g., `CONTAINER_PATCHES=/data/container_patches`

At startup you will now see output similar to this in the log:

```console
[info] Checking module compatibility with FoundryVTT version 14.360
[info] Installing semver package
[warn] Manifest URL not found in /data/Data/modules/geekpad-felddy/module.json
[info] ✅ Carousel Combat Tracker (5.0.0), Verified: 14
[info] 🟠 Automated Animations (6.8.2), Compatibility risk: last verified for version 13
[info] ✅ Combat Booster: Turn Marker, Recent Actions and more (6.0.0), Verified: 14
[info] 🟠 Break Time (13.03), Compatibility risk: last verified for version 13
[info] 🟠 Dice Tray (3.5.4), Compatibility risk: last verified for version 13
[info] ⚠️ Dice So Nice! (5.3.4), Compatibility risk: last verified for version 14.359
[info] ✅ Gambit's FXMaster (7.5.2), Verified: 14
[info] ❌ Pathfinder Beginner Box (4.2.0), UNSUPPORTED: maximum version: 13
[info] 🟠 libWrapper (1.13.4.0), Compatibility risk: last verified for version 13
[info] 🟠 PF2e Pathmuncher: Pathbuilder character importer (1.6.0), Compatibility risk: last verified for version 13
[info] ❌ PF2e HUD (2.44.3), UNSUPPORTED: maximum version: 13
[info] 🟠 pf2e award xp (1.0.8), Compatibility risk: last verified for version 13
[info] ✅ PF2e Dailies (4.17.0), Verified: 14.360
[info] ❌ Pathfinder Tokens: Monster Core (13.0.4), UNSUPPORTED: maximum version: 13
[info] ❌ PF2e Toolbelt (3.40.3), UNSUPPORTED: maximum version: 13
[info] 🟠 PF2e Dorako UI (4.1.6), Compatibility risk: last verified for version 13
[info] 🟠 Pf2E Modifiers Matter (1.16.4), Compatibility risk: last verified for version 13
[info] ❌ PopOut! (2.23), UNSUPPORTED: maximum version: 13
[info] ❌ Sequencer (3.6.11), UNSUPPORTED: maximum version: 13
[info] 🟠 Polyglot (2.8.2), Compatibility risk: last verified for version 13
[info] ⚠️ Quick Insert - Search Widget (3.7.3), Compatibility risk: last verified for version 14.359
[info] ✅ SmallTime (2.0.4), Verified: 14
[info] ✅ socketlib (v1.1.4), Verified: 14
[info] ✅ Splatter (6.0.0), Verified: 14
[info] 🟠 Foundry VTT Anniversary Dice Set by The Rollsmith (1.0.1), Compatibility risk: last verified for version 11
[info] 🟠 Tokenizer (5.0.3), Compatibility risk: last verified for version 13
[info] ✅ Wayfinder (14.0.0), Verified: 14
[info] ✅ PF2e Workbench (7.0.3), Verified: 14.360
[info] Completed URL patching.
```

This feature was requested by @cs96and, who also helped debug and improve it during development.

See:
- https://github.com/felddy/foundryvtt-docker/issues/979
- https://github.com/foundryvtt/foundryvtt/issues/11024
- https://github.com/felddy/foundryvtt-docker/issues/1242

**Comment by kinland** [comment]
I'm having some trouble getting this to work with 12.343.0; does this still work with v12+, or was this script v11 only? 
Obviously, the script needed the version changed, so CONTAINER_PATCH_URLS won't work here.

if I run a 13.344.0 image, configured the same way, I at least see a "Checking module compatibility with FoundryVTT" and "Installing semver package" log messages, but it doesn't provide any further log output.

Setting CONTAINER_PATCH_URLS or CONTAINER_PATCHES doesn't seem to change log output for v12 whatsoever. I tried executing it directly and got this error over and over with different line numbers:
> ./foundryvtt-module-check.sh: line 11: log: not found

If I add
```
alias log=echo
alias log_error=echo
alias log_warn=echo
```
to the script, then run it, it seems to execute.

**Comment by felddy** [comment]
I've updated the gist to check against `v14.360`.  It also now installs `semver` to the `/tmp` directory to avoid permission errors.

**Comment by alterNERDtive** [comment]
Neat. Any chance of making that a command line argument / environment variable in the future?

(Also seeing that jq stuff again makes me appreciate nushell :))
